src/store/modules/filterSubModule.js

```diff
@@ -0,0 +1,150 @@
+  }
+
+  label
+
+  slug
+
+  apiKey
+
+  apiValuesUnion = false
+
+  multiple = false
+
+  options = []
+
+  /**
+   * @returns {*[]}
+   */
+  get items() {
+    // eslint-disable-next-line no-use-before-define
+    return Array.isArray(this.options)
+      ? this.options
+      // eslint-disable-next-line no-use-before-define
+      : this.options[filterState.type?.value] || [];
+  }
+
+  /**
+   * select one or more options.
+   * @param selection string|string[]
+   */
+  select(selection) {
+    const selected = [selection].flat(); // coerce to array
+    this.items.forEach((option) => {
+      option.select(selection ? selected.includes(option.slug) : option.default);
+    });
+    return this;
+  }
+
+  get selectedOptions() {
+    return this.items.filter((option) => option.selected);
+  }
+
+  get selectedOption() {
+    if (this.multiple) throw Error(`Can't return single value for multiple select: ${this.label}`);
+    return this.items.find((option) => option.selected);
+  }
+
+  get value() {
+    return this.multiple
+      ? this.selectedOptions.map((option) => option.slug)
+      : this.selectedOption?.slug;
+  }
+}
+
+const apiValueFormatter = (x) => (x.includes('*') ? x : `"${x}"`);
+/**
+ * get array of API query entries for all filters
+ * Needs a selection mapping for the 'type' filter, because of special 'any' type.
+ * @param filterState
+ * @returns string[]
+ */
+const mapFiltersToApi = (filters) => (fileType) => Object.values({
+  ...filters,
+  type: (new Filter(filters.type)).select(fileType),
+}).flatMap((filter) => {
+  // get array of api values for the selected item(s)
+  const apiValues = filter.selectedOptions?.flatMap(({ apiValue }) => apiValue || []);
+  if (!apiValues?.length) return [];
+  return filter.apiValuesUnion
+    // if apiValuesUnion is selected, the values are joined by the 'OR' operator
+    ? `${filter.apiKey}:(${apiValues.map(apiValueFormatter).join(' OR ')})`
+    // otherwise, they are put as separate key:value pairs, (which act conjunctively)
+    : apiValues.map((apiValue) => `${filter.apiKey}:${apiValue}`);
+});
+
+const mutations = {
+  setRouteParams(state, routeParams) {
+    // map query parameters to state
+    filterDefinitions.forEach(({ slug }) => {
+      state[slug].select(routeParams[slug]);
+    });
+  },
+};
+
+const filterState = filterDefinitions.reduce((p, definition) => (
+  { [definition.slug]: new Filter(definition), ...p }), {});
+
+export default {
+  namespaced: true,
+  state: filterState,
+  mutations,
+  getters: {
+    uiFilters: (state) => filterDefinitions.map(({ slug }) => state[slug])
+      // the type filter is placed in another component than the other filters.
+      .filter(({ slug }) => slug !== 'type')
+      // exclude filters that don't have any options
+      // such as file extension filter when 'type'-filter is 'any' or 'directories'
+      .filter(({ items }) => items.length),
+    typeFilter: (state) => state.type,
+    mapFiltersToApi,
+  },
+};
```

Comment:
Perhaps it's because I've never seen or used JS classes, but I find this code extremely hard to read. Also, there are quite a few (too many!) eslint exceptions.

At the very least, could you add a larger comment describing what does does, e.g. the kind of dataformat it will expect? Preferably also deal with eslint errors. Yes, I'm a style nazi - mostly about consistency though.

Reply:
is this referring to the getters? Or the whole file?

Reply:
The whole module.